File: app/src/main/java/com/meiji/daily/util/RecyclerViewUtil.java
package com.meiji.daily.util;

import android.support.v7.widget.RecyclerView;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 * Created by Meiji on 2017/7/13.
 */

public class RecyclerViewUtil {

    /**
     * 使RecyclerView缓存中在pool中的Item失效
     */
    public static void invalidateCacheItem(RecyclerView recyclerView) {
        Class<RecyclerView> recyclerViewClass = RecyclerView.class;
        try {
            Field declaredField = recyclerViewClass.getDeclaredField("mRecycler");
            declaredField.setAccessible(true);
            Method declaredMethod = Class.forName(RecyclerView.Recycler.class.getName()).getDeclaredMethod("clear", (Class<?>[]) new Class[0]);
            declaredMethod.setAccessible(true);
            declaredMethod.invoke(declaredField.get(recyclerView));
            RecyclerView.RecycledViewPool recycledViewPool = recyclerView.getRecycledViewPool();
            recycledViewPool.clear();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/meiji/daily/util/RecyclerViewUtil.kt
package com.meiji.daily.util

import android.support.v7.widget.RecyclerView

/**
 * Created by Meiji on 2017/7/13.
 */

object RecyclerViewUtil {

    /**
     * 使RecyclerView缓存中在pool中的Item失效
     */
    fun invalidateCacheItem(recyclerView: RecyclerView) {
        val recyclerViewClass = RecyclerView::class.java
        try {
            val declaredField = recyclerViewClass.getDeclaredField("mRecycler")
            declaredField.isAccessible = true
            val declaredMethod = Class.forName(RecyclerView.Recycler::class.java.name)
                    .getDeclaredMethod("clear", *arrayOfNulls<Class<*>>(0) as Array<Class<*>>)
            declaredMethod.isAccessible = true
            declaredMethod.invoke(declaredField.get(recyclerView))
            val recycledViewPool = recyclerView.recycledViewPool
            recycledViewPool.clear()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }
}
